File: iOSApp/EP133SampleToolUITests/Robots/PadsRobot.swift
import XCTest

/// Robot for the Pads screen: group chips, the 12-pad grid, pad labels, and pad visual state.
/// XCUITest port of `PadsRobot.kt`.
///
/// Out-of-process caveat on touch: the pad grid handles input through a transparent UIKit
/// multi-touch overlay (the Android grid-level `pointerInteropFilter` analog), not per-pad tap
/// handlers. XCUITest can't hold one touch down and assert the pressed state before releasing it
/// the way the Compose `performTouchInput { down }` / `up()` split does — `press(forDuration:)`
/// completes the whole down→up synchronously. So `tapPad` drives a real touch (down+up) at the
/// pad's center, and the held-`pressed`-state assertion is adapted in the test class.
final class PadsRobot: BaseRobot {

    @discardableResult
    func selectGroup(_ group: String) -> PadsRobot {
        // The group-chip tag is shared with the Kit screen (both stay mounted); tap the visible one.
        hittableElement(TestTags.groupChip(group)).tap()
        return self
    }

    /// Tap a pad at its center — a full down+up touch that lands on the overlay touch layer,
    /// firing padDown then padUp (the `tapPad` analog).
    @discardableResult
    func tapPad(_ index: Int) -> PadsRobot {
        element(TestTags.pad(index)).tap()
        return self
    }

    /// Press-and-hold a pad for [duration]. The overlay fires padDown on touch-begin and padUp on
    /// touch-end; the pressed glow is held only while down, so this is the closest out-of-process
    /// analog of the Compose down/hold/up.
    @discardableResult
    func pressPad(_ index: Int, forDuration duration: TimeInterval = 0.3) -> PadsRobot {
        element(TestTags.pad(index)).press(forDuration: duration)
        return self
    }

    /// Assert a pad's visual state via its accessibilityValue: "pressed" / "in_scale" / "idle"
    /// (the Android `stateDescription` analog).
    @discardableResult
    func assertPadState(_ index: Int, _ state: String) -> PadsRobot {
        let pad = waitForTag(TestTags.pad(index))
        XCTAssertEqual(
            pad.value as? String, state,
            "pad \(index) should be in state '\(state)' (was '\(String(describing: pad.value))')")
        return self
    }

    @discardableResult
    func assertPadLabel(_ label: String) -> PadsRobot {
        assertTextDisplayed(label)
        return self
    }

    @discardableResult
    func assertGridVisible() -> PadsRobot {
        assertTagDisplayed(TestTags.PADS_GRID)
        return self
    }
}

File: AndroidApp/app/src/androidTest/java/com/ep133/sampletool/robots/PadsRobot.kt
package com.ep133.sampletool.robots

import androidx.compose.ui.test.assert
import androidx.compose.ui.test.assertIsDisplayed
import androidx.compose.ui.test.junit4.ComposeContentTestRule
import androidx.compose.ui.test.performClick
import androidx.compose.ui.test.performTouchInput
import com.ep133.sampletool.ui.TestTags

/** Robot for the Pads screen: group chips, the 12-pad grid, and pad visual states. */
class PadsRobot(rule: ComposeContentTestRule) : BaseRobot(rule) {

    fun selectGroup(group: String) = apply {
        tag(TestTags.groupChip(group)).performClick()
    }

    /**
     * Touch-down on a pad without releasing. The grid handles touch via a parent
     * pointerInteropFilter, so this injects at the pad cell's center and lets the
     * parent's coordinate math resolve the index — the same path a finger takes.
     */
    fun pressPad(index: Int) = apply {
        tag(TestTags.pad(index)).performTouchInput { down(center) }
    }

    fun releasePad(index: Int) = apply {
        tag(TestTags.pad(index)).performTouchInput { up() }
    }

    fun tapPad(index: Int) = apply {
        tag(TestTags.pad(index)).performTouchInput {
            down(center)
            up()
        }
    }

    /** Assert a pad's visual state via its stateDescription: "pressed" / "in_scale" / "idle". */
    fun assertPadState(index: Int, state: String) = apply {
        tag(TestTags.pad(index)).assert(hasStateDescription(state))
    }

    fun assertPadLabel(label: String) = apply {
        text(label).assertIsDisplayed()
    }

    fun assertGridVisible() = apply {
        tag(TestTags.PADS_GRID).assertIsDisplayed()
    }
}
